src/coprocessor/coprocessor.rs

```diff
@@ -0,0 +1,1142 @@
+                    ranges.as_slice(),
+                    self.max_handle_duration,
+                    peer,
+                    Some(is_desc_scan),
+                    Some(dag.get_start_ts()),
+                );
+                let batch_row_limit = self.get_batch_row_limit(is_streaming);
+                builder = box move |snap, req_ctx: &_| {
+                    // See rust-lang#41078 to know why we have `: &_` here.
+                    dag::DAGContext::new(dag, ranges, snap, req_ctx, batch_row_limit)
+                        .map(|h| h.into_boxed())
+                };
+            }
+            REQ_TYPE_ANALYZE => {
+                let mut analyze = AnalyzeReq::new();
+                box_try!(analyze.merge_from(&mut is));
+                let table_scan = analyze.get_tp() == AnalyzeType::TypeColumn;
+                req_ctx = ReqContext::new(
+                    make_tag(table_scan),
+                    context,
+                    ranges.as_slice(),
+                    self.max_handle_duration,
+                    peer,
+                    None,
+                    Some(analyze.get_start_ts()),
+                );
+                builder = box move |snap, req_ctx: &_| {
+                    statistics::analyze::AnalyzeContext::new(analyze, ranges, snap, req_ctx)
+                        .map(|h| h.into_boxed())
+                };
+            }
+            REQ_TYPE_CHECKSUM => {
+                let mut checksum = ChecksumRequest::new();
+                box_try!(checksum.merge_from(&mut is));
+                let table_scan = checksum.get_scan_on() == ChecksumScanOn::Table;
+                req_ctx = ReqContext::new(
+                    make_tag(table_scan),
+                    context,
+                    ranges.as_slice(),
+                    self.max_handle_duration,
+                    peer,
+                    None,
+                    Some(checksum.get_start_ts()),
+                );
+                builder = box move |snap, req_ctx: &_| {
+                    checksum::ChecksumContext::new(checksum, ranges, snap, req_ctx)
+                        .map(|h| h.into_boxed())
+                };
+            }
+            tp => return Err(box_err!("unsupported tp {}", tp)),
+        };
+        Ok((builder, req_ctx))
+    }
+
+    /// Parse the raw `Request` to create `RequestHandlerBuilder` and `ReqContext`.
+    #[inline]
+    fn parse_request(
+        &self,
+        req: coppb::Request,
+        peer: Option<String>,
+        is_streaming: bool,
+    ) -> (RequestHandlerBuilder<E::Snap>, ReqContext) {
+        match self.try_parse_request(req, peer, is_streaming) {
+            Ok(v) => v,
+            Err(err) => {
+                // If there are errors when parsing requests, create a dummy request handler.
+                let builder =
+                    box |_, _: &_| Ok(cop_util::ErrorRequestHandler::new(err).into_boxed());
+                let req_ctx = ReqContext::new(
```

Comment:
oh, so we need to get the async snapshot for the invalid request and send this to execute in the read pool too?

Reply:
yes. However invalid requests should be rare, so the cost is affordable.